mpack-instance-manager/src/test/python/instance_manager/test_instance_manager.py

```diff
@@ -0,0 +1,265 @@
+                    "hdfs": {
+                      "category": "SERVER",
+                      "name": "hdfs",
+                      "components": {
+                        "hdfs_server": {
+                          "component-instances": {
+                            "server1": {
+                              "config_dir": "/tmp/instance_manager_test/instances/hdpcore/Production/default/hdfs/hdfs_server/server1/conf",
+                              "name": "server1"
+                            }
+                          }
+                        }
+                      }
+                    },
+                    "hdfs_client": {
+                      "category": "CLIENT",
+                      "component_instances": {
+                        "default": {
+                          "config_dir": "/tmp/instance_manager_test/instances/hdpcore/Production/default/hdfs_client/conf",
+                          "name": "default"
+                        }
+                      },
+                      "name": "hdfs_client"
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    self.assertEqual(conf_dir_json, expected_json)
+
+  def test_list_instances_all(self):
+    create_mpack_with_defaults(module_name=CLIENT_MODULE_NAME)
+    create_mpack_with_defaults(module_name=SERVER_MODULE_NAME, components=None,
+                               components_map={SERVER_COMPONENT_NAME: ['server1']})
+
+    conf_dir_json = instance_manager.list_instances()
+
+    expected_json = {
+      "mpacks": {
+        "hdpcore": {
+          "mpack-instances": {
+            "Production": {
+              "name": "Production",
+              "subgroups": {
+                "default": {
+                  "modules": {
+                    "hdfs": {
+                      "category": "SERVER",
+                      "name": "hdfs",
+                      "components": {
+                        "hdfs_server": {
+                          "component-instances": {
+                            "server1": {
+                              "path": "/tmp/instance_manager_test/modules/hdfs/3.1.0.0-b1",
+                              "name": "server1"
+                            }
+                          }
+                        }
+                      }
+                    },
+                    "hdfs_client": {
+                      "category": "CLIENT",
+                      "component_instances": {
+                        "default": {
+                          "path": "/tmp/instance_manager_test/modules/hdfs-clients/3.1.0.0-b1",
+                          "name": "default"
+                        }
+                      },
+                      "name": "hdfs_client"
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    self.assertEqual(conf_dir_json, expected_json)
+
```

Comment:
Thanks for covering the tests part. Can we also have some -ve tests also ? 
Few examples: 
 - CREATE : if some component instance already exists ?
 - CREATE : 2nd create call to add an instance(s) so as to build up on the already existing structure of mpack/subgroup/instance? 
 - SET : for an mpack/subgroup/module/instance which doesn't exists?
 - LIST/CONF : Something (component instance) at terminal level which doesnt exist?

Reply:
Added negative case scenarious